QUARKUS-1094.md

```diff
@@ -0,0 +1,24 @@
+# QUARKUS-1094 - Initial support of the gRPC in the dev ui
+
+JIRA link: https://issues.redhat.com/browse/QUARKUS-1094
+
+Quarkus has an experimental feature for gRPC in the DevUI. Now it is possible to test gRPC services provided by the server without implementing clients. The gRPC uses HTTP2 which is well suited for microservices inner-communication, but there is no native way for browser client to utilize HTTP/2, thus the feature utilizes Web Socket Bridge for communication.
```

Comment:
I don't understand the last sentence `there is no native way for browser client to utilize HTTP/2...`.

Reply:
Even though browsers have support for HTTP2, the browser JS doesn't have an API to work with binary data frames (because everything in HTTP2 is binary). So there are multiple solutions, the first one to use gRPC proxy (https://grpc.io/blog/state-of-grpc-web/#the-tech) and the second one, which DevUI is using, is utilizing WebSockets. The gRPC request/response can be transcoded to WebSockets (according to https://cloud.redhat.com/blog/grpc-anywhere), but WebSockets utilizing HTTP/1.1. Under the hood when using DevUI with gRPC there is a connection upgrade to `Upgrade: websocket`, because of that I don't think `htmlunit` can handle the test. 
From the above, I don't see an automation solution on HTTP level, and because the feature implements WebSockets on the server-side, the option for the test could be**:** WebSockets client that connects to WebSockets server and tests proper transcoding from gRPC to WebSockets.

P.S that's how I understood the problem, you can challenge this opinion of course

Reply:
Please add details from this comment (gRPC proxy + websockets way) into TP to provide additional context. 

About `htmlunit` and HTTP upgrade .. are you saying that you may not be able to automate the testing using current tool chain?

Reply:
I found some `htmlunit support for websockets`, but this should be verified. There are no really guides about that. https://sourceforge.net/p/htmlunit/code/13880/tree/trunk/htmlunit/src/test/java/com/gargoylesoftware/htmlunit/javascript/host/WebSocketTest.java#l484

Reply:
HtmlUnit can not handle javascript execution because of its JS engine. TD implementation will be based on OkHttp library with its support for websockets